### ISSUE TITLE
Bouncy Castle exception

### ISSUE THREAD

**Issue by macourtney** [macourtney/Dark-Exchange#1]
java.security.NoSuchAlgorithmException: Error constructing implementation (algorithm: SHA1WithRSAEncryption, provider: BC, class: org.bouncycastle.jce.provider.JDKDigestSignature$SHA1WithRSAEncryption)
    at java.security.Provider$Service.newInstance(Provider.java:1245)
    at sun.security.jca.GetInstance.getInstance(GetInstance.java:220)
    at sun.security.jca.GetInstance.getInstance(GetInstance.java:190)
    at java.security.Signature.getInstance(Signature.java:324)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
    at java.lang.reflect.Method.invoke(Method.java:597)
    at clojure.lang.Reflector.invokeMatchingMethod(Reflector.java:90)
    at clojure.lang.Reflector.invokeStaticMethod(Reflector.java:202)
    at darkexchange.model.security$sign.invoke(security.clj:139)
    at darkexchange.model.user$sign.invoke(user.clj:141)
    at darkexchange.model.interceptors.signature_interceptor$sign.invoke(signature_interceptor.clj:17)
    at darkexchange.model.interceptors.signature_interceptor$client_interceptor.invoke(signature_interceptor.clj:47)
    at darkexchange.model.interceptors.interceptor_util$compose_interceptors$fn__4707$fn__4708.invoke(interceptor_util.clj:4)
    at darkexchange.model.interceptors.identity_interceptor$client_interceptor.invoke(identity_interceptor.clj:17)
    at darkexchange.model.interceptors.interceptor_util$compose_interceptors$fn__4707.invoke(interceptor_util.clj:4)
    at darkexchange.model.interceptors.interceptor_util$run_interceptors.invoke(interceptor_util.clj:9)
    at darkexchange.model.interceptors.client_interceptors$run_interceptors.invoke(client_interceptors.clj:10)
    at darkexchange.model.client$send_message.invoke(client.clj:26)
    at darkexchange.model.peer$get_peers_from.invoke(peer.clj:98)
    at darkexchange.model.peer$all_network_destinations.invoke(peer.clj:152)
    at darkexchange.model.peer$download_peers_background.invoke(peer.clj:159)
    at darkexchange.model.peer$download_peers_background.invoke(peer.clj:164)
    at darkexchange.model.peer$download_peers_background.invoke(peer.clj:155)
    at clojure.lang.AFn.run(AFn.java:24)
    at java.lang.Thread.run(Thread.java:662)
Caused by: java.lang.SecurityException: class "org.bouncycastle.crypto.Digest"'s signer information does not match signer information of other classes in the same package
    at java.lang.ClassLoader.checkCerts(ClassLoader.java:807)
    at java.lang.ClassLoader.preDefineClass(ClassLoader.java:488)
    at java.lang.ClassLoader.defineClassCond(ClassLoader.java:626)
    at java.lang.ClassLoader.defineClass(ClassLoader.java:616)
    at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:141)
    at java.net.URLClassLoader.defineClass(URLClassLoader.java:283)
    at java.net.URLClassLoader.access$000(URLClassLoader.java:58)
    at java.net.URLClassLoader$1.run(URLClassLoader.java:197)
    at java.security.AccessController.doPrivileged(Native Method)
    at java.net.URLClassLoader.findClass(URLClassLoader.java:190)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:307)
    at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:301)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:248)
    at java.lang.Class.getDeclaredConstructors0(Native Method)
    at java.lang.Class.privateGetDeclaredConstructors(Class.java:2389)
    at java.lang.Class.getConstructor0(Class.java:2699)
    at java.lang.Class.newInstance0(Class.java:326)
    at java.lang.Class.newInstance(Class.java:308)
    at java.security.Provider$Service.newInstance(Provider.java:1221)
    ... 26 more
ERROR [darkexchange.model.interceptors.signature-interceptor]: an error occured while signing or verifying a signature on the client side.
java.security.NoSuchAlgorithmException: Error constructing implementation (algorithm: SHA1WithRSAEncryption, provider: BC, class: org.bouncycastle.jce.provider.JDKDigestSignature$SHA1WithRSAEncryption)
    at java.security.Provider$Service.newInstance(Provider.java:1245)
    at sun.security.jca.GetInstance.getInstance(GetInstance.java:220)
    at sun.security.jca.GetInstance.getInstance(GetInstance.java:190)
    at java.security.Signature.getInstance(Signature.java:324)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
    at java.lang.reflect.Method.invoke(Method.java:597)
    at clojure.lang.Reflector.invokeMatchingMethod(Reflector.java:90)
    at clojure.lang.Reflector.invokeStaticMethod(Reflector.java:202)
    at darkexchange.model.security$sign.invoke(security.clj:139)
    at darkexchange.model.user$sign.invoke(user.clj:141)
    at darkexchange.model.interceptors.signature_interceptor$sign.invoke(signature_interceptor.clj:17)
    at darkexchange.model.interceptors.signature_interceptor$client_interceptor.invoke(signature_interceptor.clj:47)
    at darkexchange.model.interceptors.interceptor_util$compose_interceptors$fn__4707$fn__4708.invoke(interceptor_util.clj:4)
    at darkexchange.model.interceptors.identity_interceptor$client_interceptor.invoke(identity_interceptor.clj:17)
    at darkexchange.model.interceptors.interceptor_util$compose_interceptors$fn__4707.invoke(interceptor_util.clj:4)
    at darkexchange.model.interceptors.interceptor_util$run_interceptors.invoke(interceptor_util.clj:9)
    at darkexchange.model.interceptors.client_interceptors$run_interceptors.invoke(client_interceptors.clj:10)
    at darkexchange.model.client$send_message.invoke(client.clj:26)
    at darkexchange.model.calls.notify$call.invoke(notify.clj:10)
    at darkexchange.model.peer$notify_destination.invoke(peer.clj:92)
    at darkexchange.model.peer$notify_all_peers.invoke(peer.clj:134)
    at darkexchange.model.peer$add_destinations.invoke(peer.clj:147)
    at darkexchange.model.peer$download_peers_background.invoke(peer.clj:161)
    at darkexchange.model.peer$download_peers_background.invoke(peer.clj:164)
    at darkexchange.model.peer$download_peers_background.invoke(peer.clj:155)
    at clojure.lang.AFn.run(AFn.java:24)
    at java.lang.Thread.run(Thread.java:662)
Caused by: java.lang.SecurityException: class "org.bouncycastle.crypto.Digest"'s signer information does not match signer information of other classes in the same package
    at java.lang.ClassLoader.checkCerts(ClassLoader.java:807)
    at java.lang.ClassLoader.preDefineClass(ClassLoader.java:488)
    at java.lang.ClassLoader.defineClassCond(ClassLoader.java:626)
    at java.lang.ClassLoader.defineClass(ClassLoader.java:616)
    at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:141)
    at java.net.URLClassLoader.defineClass(URLClassLoader.java:283)
    at java.net.URLClassLoader.access$000(URLClassLoader.java:58)
    at java.net.URLClassLoader$1.run(URLClassLoader.java:197)
    at java.security.AccessController.doPrivileged(Native Method)
    at java.net.URLClassLoader.findClass(URLClassLoader.java:190)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:307)
    at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:301)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:248)
    at java.lang.Class.getDeclaredConstructors0(Native Method)
    at java.lang.Class.privateGetDeclaredConstructors(Class.java:2389)
    at java.lang.Class.getConstructor0(Class.java:2699)
    at java.lang.Class.newInstance0(Class.java:326)
    at java.lang.Class.newInstance(Class.java:308)
    at java.security.Provider$Service.newInstance(Provider.java:1221)
    ... 28 more

**Comment by macourtney** [comment]
Notes from Steve:

I have not seen this error message before. However, a web search indicates that its likely that bouncy castle class files are being loaded from different jars - that are not compatible with each other.  I inspected a couple of jars in the lib directory and found that ...
i2p-0.8.5-0.jar contains org.bouncycastle.crypto.Digest.class
bcprov-jdk16-136.jar contains org.bouncycastle.crypto.Digest.class

I attempted to confirm the cause by executing Java with the -verbose option to record the jars from which the org.bouncycastle.crypto classes are loaded. I found that they are all loaded from bcprov-jdk16-136.jar.  I tried removing bcprov-jdk16-136.jar from the lib directory, and the application fails to launch because it cannot find org.bouncycastle.jce.provider.BouncyCastleProvider.

I tried upgrading bcprov-jdk16-136.jar to bcprov-jdk16-140.jar and same problem.

I explicitly set the class path jar file order to ...cp=lib/clj-record-1.0.1.jar
cp=$cp:lib/clojure-1.2.1.jar
cp=$cp:lib/clojure-contrib-1.2.0.jar
cp=$cp:lib/clojure-tools-1.0.0.jar
cp=$cp:lib/commons-codec-1.5.jar
cp=$cp:lib/commons-lang-2.5.jar
cp=$cp:lib/darkexchange-1.0.0-BETA.jar
cp=$cp:lib/drift-1.2.1.jar
cp=$cp:lib/h2-1.2.137.jar
cp=$cp:lib/i2p-0.8.5-0.jar
cp=$cp:lib/log4j-1.2.16.jar
cp=$cp:lib/miglayout-3.7.4.jar
cp=$cp:lib/mstreaming-0.8.5-0.jar
cp=$cp:lib/seesaw-1.0.6.jar
cp=$cp:lib/streaming-0.8.5-0.jar
cp=$cp:lib/bcprov-jdk16-140.jar
cp=$cp:resources

But the same problem occurred.

Then I removed the org.bouncycastle package from i2p-0.8.5-0.jar with the Ubuntu File Roller graphical utility, and the problem went away.  The search feature returns no results for the couple of choices that I tried, but there was no security exception.

Here is the relevant part of the log ...
INFO  [config.db-config]: Environment: :production
DEBUG [darkexchange.database.h2]: Executing query: ["SELECT \* FROM schema_info LIMIT 1"]
INFO  [darkexchange.database.migrate]: schema_info exists
DEBUG [darkexchange.database.h2]: Executing query: ["SELECT \* FROM schema_info LIMIT 1"]
INFO  [drift.runner]: Current database version: 7
INFO  [drift.runner]: Updating to version: 7
INFO  [drift.runner]: No changes were made to the database.
INFO  [darkexchange.model.server]: Initializing server.
INFO  [darkexchange.model.listeners.listener-init]: Adding listeners.
INFO  [darkexchange.model.actions.action-init]: Adding actions.

Hope this is helpful.
-Steve
